sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java

```diff
@@ -0,0 +1,228 @@
+
+import javax.annotation.Nullable;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import static org.hamcrest.Matchers.is;
+
+public class IngestionServiceIntegrationTest extends SQLTransportIntegrationTest {
+
+    private static final String INGESTION_SOURCE_ID = "testing";
+    private static final String UNDER_TEST_INGEST_RULE_NAME = "ingest_data_rule";
+    private static final String DATA_TOPIC = "test";
+    private static final String TARGET_TABLE = "ingest_data_raw";
+
+    private class TestIngestionImplementation implements IngestionImplementation {
+
+        private final IngestionService ingestionService;
+        private final InputFactory inputFactory;
+        private Set<IngestRule> ingestRules;
+        private final ExpressionAnalyzer expressionAnalyzer;
+        private final ExpressionAnalysisContext expressionAnalysisContext;
+        private FieldProvider<Symbol> inputColumnProvider = new FieldProvider<Symbol>() {
+
+            @Override
+            public Symbol resolveField(QualifiedName qualifiedName, Operation operation) {
+                return resolveField(qualifiedName, null, operation);
+            }
+
+            @Override
+            public Symbol resolveField(QualifiedName qualifiedName, @Nullable List<String> path, Operation operation) {
+                return new InputColumn(0);
+            }
+        };
+
+        TestIngestionImplementation(Functions functions,
+                                    IngestionService ingestionService) {
+            this.ingestRules = new HashSet<>();
+            this.inputFactory = new InputFactory(functions);
+            this.ingestionService = ingestionService;
+            this.expressionAnalysisContext = new ExpressionAnalysisContext();
+            this.expressionAnalyzer = new ExpressionAnalyzer(functions, null, null, inputColumnProvider, null);
+        }
+
+        void registerIngestionImplementation() {
+            ingestionService.registerImplementation(INGESTION_SOURCE_ID, this);
+        }
+
+        void ingestData(String topic, int data) {
+            if (ingestRules.size() > 0) {
+                for (IngestRule ingestRule : ingestRules) {
+                    Symbol conditionSymbol = expressionAnalyzer.convert(SqlParser.createExpression(ingestRule.getCondition()), expressionAnalysisContext);
+                    Predicate<Row> conditionPredicate = RowFilter.create(inputFactory, conditionSymbol);
+                    if (conditionPredicate.test(new Row1(topic))) {
+                        execute(String.format(Locale.ENGLISH,
+                            "insert into %s values (%s);", ingestRule.getTargetTable(), data)
```

Comment:
`data` should never be passed in using string format. even that this is a test, I wouldn't do it ;)